src/muz/spacer/spacer_unsat_core_learner.cpp

```diff
@@ -36,28 +36,27 @@ void unsat_core_learner::register_plugin(unsat_core_plugin* plugin) {
 }
 
 void unsat_core_learner::compute_unsat_core(expr_ref_vector& unsat_core) {
-    // traverse proof
     proof_post_order it(m_pr.get(), m);
     while (it.hasNext()) {
-        proof* currentNode = it.next();
+        proof* curr = it.next();
 
-        if (m.get_num_parents(currentNode) > 0) {
-            bool need_to_mark_closed = true;
+        bool done = is_closed(curr);
+        if (done) continue;
 
-            for (proof* premise : m.get_parents(currentNode)) {
-                need_to_mark_closed &= (!m_pr.is_b_marked(premise) || m_closed.is_marked(premise));
-            }
-
-            // save result
-            m_closed.mark(currentNode, need_to_mark_closed);
+        if (m.get_num_parents(curr) > 0) {
+            done = true;
+            for (proof* p : m.get_parents(curr)) done &= !is_b_open(p);
+            set_closed(curr, done);
         }
 
-        // we have now collected all necessary information, so we can visit the node
-        // if the node mixes A-reasoning and B-reasoning and contains non-closed premises
-        if (m_pr.is_a_marked(currentNode) && 
-            m_pr.is_b_marked(currentNode) && 
-            !m_closed.is_marked(currentNode)) {
-            compute_partial_core(currentNode); // then we need to compute a partial core
+        // we have now collected all necessary information,
+        // so we can visit the node
+        // if the node mixes A-reasoning and B-reasoning
+        // and contains non-closed premises
+        if (!done) {
```

Comment:
always true

Reply:
@NikolajBjorner why? `done` is true if all parents are `done`. This is entered only if there is an open parent that has not been interpolated over.